### PR TITLE
Use `js-sys` regex type on wasm32-unknown-unknown with feature enabled

### PR DESCRIPTION
**Pull request by aumetra**
This is still missing a workflow to run the tests in CI but they should work

Edit: `insta` just [doesn't work with `wasm-bindgen-test`](https://github.com/mitsuhiko/insta/issues/120), so testing is not really feasible with the current test setup

### CODE REVIEW

**Comment by aumetra** [comment]
Hacked together a small regex test that doesn't use the `insta` snapshots.  
It's not pretty but it runs!

File: garde/src/rules/pattern.rs [hunk 1]
```diff
@@ -69,6 +69,59 @@ impl<T: Pattern> Pattern for Option<T> {
     }
 }
 
+#[cfg(all(
+    feature = "regex",
+    all(feature = "js-sys", all(target_arch = "wasm32", target_os = "unknown"))
+))]
```
Comment (jprochazk) [inline garde/src/rules/pattern.rs:75]: what's with all the nesting? (no pun intended) isnt `all` just `&&`? so you could:
```suggestion
#[cfg(all(
    feature = "regex",
    feature = "js-sys",
    target_arch = "wasm32",
    target_os = "unknown",
))]
```

File: garde/src/rules/pattern.rs [hunk 2]
```diff
@@ -69,6 +69,59 @@ impl<T: Pattern> Pattern for Option<T> {
...
+
+    impl RegExpInvoker {
+        pub const fn new(pattern: &'static str) -> Self {
+            Self(pattern)
+        }
+    }
+
+    impl Matcher for RegExpInvoker {
+        fn is_match(&self, haystack: &str) -> bool {
+            RegExp::new(self.0, "u").is_match(haystack)
```
Comment (jprochazk) [inline garde/src/rules/pattern.rs:104]: constructing a new regex each time doesn't seem great, can this be cached somehow?

File: garde/src/rules/pattern.rs (hunk 1 above)
Comment (jprochazk) [reply garde/src/rules/pattern.rs:75]: same deal in other places

**Comment by jprochazk** [comment]
i know it's a draft, just threw in some drive-by comments

File: garde/src/rules/pattern.rs (hunk 2 above)
Comment (aumetra) [reply garde/src/rules/pattern.rs:104]: I tried. The issue here is that the `js-sys` types are inherently not `Sync`, so we can't really store them inside a `static`. It's a pain

Comment (jprochazk) [reply garde/src/rules/pattern.rs:104]: yeah but wasm is not multithreaded, so it's fine to use a bit of unsafe code to put it in a static

Comment (jprochazk) [reply garde/src/rules/pattern.rs:104]: specifically `wasm32-unknown-unknown` will forever be single-threaded, threading is enabled via a different target (one of the `wasi` ones, i dont remember which exactly)

File: garde/src/rules/pattern.rs (hunk 1 above)
Comment (aumetra) [reply garde/src/rules/pattern.rs]: Flattened all the expressions

File: garde/src/rules/pattern.rs (hunk 2 above)
Comment (aumetra) [reply garde/src/rules/pattern.rs]: Added an unsafe `SyncWrapper` type that can't really be constructed outside of the crate without usage of `unsafe` with the explicit disclaimer that either the value must never be passed or shared between threads, or the inner type has to be `Send` and `Sync`

**Comment by aumetra** [comment]
So, it's now looking more cleaned up, the regex is cached between invocations, and there is a new CI step

**Review by jprochazk — APPROVED** [review]

**Pull request merged**